src/ray/common/bundle_spec.cc

```diff
@@ -90,12 +94,22 @@ std::string FormatPlacementGroupResource(const std::string &original_resource_na
                                          const PlacementGroupID &group_id,
                                          int64_t bundle_index) {
   std::string str;
+  const auto &group_id_hex = group_id.Hex();
   if (bundle_index >= 0) {
-    str = original_resource_name + "_group_" + std::to_string(bundle_index) + "_" +
-          group_id.Hex();
+    const auto &bundle_index_str = std::to_string(bundle_index);
```

Comment:
According to stackoverflow; https://stackoverflow.com/questions/18892281/most-optimized-way-of-concatenation-in-strings/18892355

Cpp strings are mutable and it calls `reserve` internally for concatenation which is expensive. Reserving the space earlier and appending strings using += is the most optimized approach. See https://github.com/ray-project/ray/issues/18651 for the profiling result

Reply:
How about using std::stringstream?

Reply:
According to the benchmark in the stackoverflow page, std::stringstream performance is not that great due to its formatting related stuff.

<img width="807" alt="Screen Shot 2021-09-20 at 9 38 18 AM" src="https://user-images.githubusercontent.com/18510752/134040322-9c2d10d6-a6fe-425d-8930-b7a116011120.png">

Reply:
I wouldn't trust that answer, it's pretty poorly benchmarked. Have you tried either just using `+=` operator or stringstream, both of which should be amoritzed O(n) time?

Reply:
Let me try running it myself too. Also based on additional research, it seems like stringstream performance is not too bad after C++11 (but there's still a bit of overhead compared to reserve + append because the object itself is more complex), and depending on compiler, it has pretty good performance. Given the most of overhead was from "reserve" (not other parts), I think I can just use stringstream in this case (maybe a bit of overhead in construction is negligible + stringstream will have better readability)

Reply:
How about just use https://abseil.io/docs/cpp/guides/format#advanced here if it's critical in performance.